### PR TITLE
Keysight B1500 Parameter Analyzer: Sampling measurement mode

### PR DESCRIPTION
**Pull request by astafan8**
Implement Sampling measurement mode for Keysight B1500 Semiconductor Parameter Analyzer and its B1517A SMU (Source/Monitor Unit) module. This is very useful, for example, for measuring voltage noise in the pinchoff region for a device and calculate its power spectral density.

Includes:
- [x] `ParameterWithSetpoints` that represents the sampling measurement
- [x] parser for basic ASCII `FMT 1`
- [x] ~parser for binary format that contains timestamps of the measurement points~ ("won't do" because parsing that binary format is not worth the effort)
- [x] document Sampling mode measurement in the example notebook
- [x] `B1517A.timing_parameters` method corresponding to `MT` FLEX command for settings timing parameters of sampling measurement
- [x] `B1517A.measurement_mode` parameter that wraps `MM` FLEX command
- [x] `B1500.use_manual_mode_for_high_speed_adc` convenience function that is useful in the context of sampling mode measurement
- [x] docstrings and tests for some relevant commands represented as methods of `MessageBuilder`

### CODE REVIEW

File: qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500_SamplingMeasurement.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
+import warnings
+
```
Comment (astafan8) [inline qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500_SamplingMeasurement.py]: let's rename this module to `sampling_measurement` so that it's pep8.

File: qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1517A.py [hunk 2]
```diff
@@ -49,6 +79,42 @@ def __init__(self, parent: 'KeysightB1500', name: Optional[str], slot_nr,
...
         )
 
+        self.add_parameter(
+            name="time_axis",
+            get_cmd=self._get_time_axis,
+            vals=Arrays(shape=(self._get_sampling_number,))
+        )
+
+        self.add_parameter(
+            name="sampling_measurement",
```
Comment (astafan8) [inline qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1517A.py]: i'd suggest to call this parameter `sampling_measurement_trace` or smth, not just `sampling_measurement` as it is too abstract and doesn't really convey that it's a parameter.

**Comment by WilliamHPNielsen** [comment]
As far as I can see, this is one the instruments that never called `connect_message()` in `__init__`. It would be fine to add that here. Possibly you could cherry-pick that commit to a different branch and make a quickfix PR to address that issue.

File: qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500_SamplingMeasurement.py (hunk 1 above)
Comment (lakhotiaharshit) [reply qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500_SamplingMeasurement.py]: Done

File: qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500_SamplingMeasurement.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    pass
+
+
+class SamplingMeasurement(ParameterWithSetpoints):
+    """
+    Performs sampling measurement using semiconductor
+    parameter analyzer B1500A.
+    """
+
+    _timeout_response_factor = 10.00
```
[the remark this answers is not in the document]
Comment (lakhotiaharshit) [reply qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500_SamplingMeasurement.py]: I have tested with 10. I hope 2-3 will work but I am not sure about this so at the moment I will keep it to 10 and later on once this becomes routine in the lab, we can optimize the response time.

File: qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500_SamplingMeasurement.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    # This factor is a bit higher than the ratio between
+    # the measured measurement-time and the calculated measurement
+    # (from the user input). Check :get_raw: method to find its usage.
+
+    def __init__(self, name, **kwargs):
+        super().__init__(name, **kwargs)
+        self.data = FMTResponse
```
[the remark this answers is not in the document]
Comment (lakhotiaharshit) [reply qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500_SamplingMeasurement.py]: Yes

File: qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500_SamplingMeasurement.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+    def _set_up(self):
```
[the remark this answers is not in the document]
Comment (lakhotiaharshit) [reply qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500_SamplingMeasurement.py]: done

File: qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500_SamplingMeasurement.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        the sampling measurement. The reason is that while measurement is
+        running the data keeps on appending in the buffer of SPA. Only when
+        the measurement is finished the data is returned to the VISA handle.
+        Hence during this time the VISA is idle and waiting for the response.
+        If the timeout is lower than the total run time of the measurement,
+        VISA will give error.
+
+        We set the Visa timeout to be the measurement_time times the
+        `_timeout_response_factor`. Strictly speaking the timeout should be
+        just higher the measurement time.
```
[the remark this answers is not in the document]
Comment (lakhotiaharshit) [reply qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500_SamplingMeasurement.py]: Thanks

File: qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500_SamplingMeasurement.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+            exception_count = total_count - normal_count
+            if total_count == normal_count:
+                print('All measurements are normal')
+            else:
+                indices = [i for i, x in enumerate(data.status)
+                           if x == "C" or x == "T"]
+                warnings.warn(f'{str(exception_count)} measurements were '
+                              f'out of compliance at {str(indices)}')
+
+            compliance_error_list = constants.ComplianceErrorList
```
[the remark this answers is not in the document]
Comment (lakhotiaharshit) [reply qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500_SamplingMeasurement.py]: Alias removed

File: qcodes/tests/drivers/keysight_b1500/b1500_driver_tests/test_sampling_measurement.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+        smu_sampling_measurement.sampling_measurement.compliance(), list)
+    np.testing.assert_array_equal(
+        smu_sampling_measurement.sampling_measurement.compliance(),
+        compliance_list
+    )
+
+
+def test_output_data_type_and_data_channel(smu_sampling_measurement,
+                                           smu_output):
+    n_samples = smu_output[0]
```
[the remark this answers is not in the document]
Comment (lakhotiaharshit) [reply qcodes/tests/drivers/keysight_b1500/b1500_driver_tests/test_sampling_measurement.py]: Done

File: qcodes/tests/drivers/keysight_b1500/b1500_driver_tests/test_sampling_measurement.py [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+def test_compliance_needs_data_from_sampling_measurement(smu):
+    with pytest.raises(Exception,
+                       match='First run sampling_measurement method '
+                             'to generate the data'):
+        smu.sampling_measurement.compliance()
+
+
+def test_compliance(smu_sampling_measurement,
+                    smu_output):
+    n_samples = smu_output[0]
```
[the remark this answers is not in the document]
Comment (lakhotiaharshit) [reply qcodes/tests/drivers/keysight_b1500/b1500_driver_tests/test_sampling_measurement.py]: Done

File: qcodes/tests/drivers/keysight_b1500/b1500_driver_tests/test_sampling_measurement.py [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+def test_measurement_requires_timing_parameters_to_be_set(smu):
+    with pytest.raises(Exception, match='set timing parameters first'):
+        smu.sampling_measurement.get()
+
+
+def test_sampling_measurement(smu_sampling_measurement,
+                              smu_output):
+    smu_sampling_measurement, _, _, _ = smu_sampling_measurement
+    n_samples = smu_output[0]
+    data_to_return = smu_output[1]
```
[the remark this answers is not in the document]
Comment (lakhotiaharshit) [reply qcodes/tests/drivers/keysight_b1500/b1500_driver_tests/test_sampling_measurement.py]: Done

File: qcodes/tests/drivers/keysight_b1500/b1500_driver_tests/test_sampling_measurement.py [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+
+@pytest.fixture
+def smu(b1500):
+    return b1500.smu1
+
+
+@pytest.fixture
+def smu_output():
+    n_samples = 7
+    data_to_return = np.random.rand(n_samples)
```
[the remark this answers is not in the document]
Comment (lakhotiaharshit) [reply qcodes/tests/drivers/keysight_b1500/b1500_driver_tests/test_sampling_measurement.py]: Done

File: qcodes/instrument_drivers/Keysight/keysightb1500/constants.py [hunk 12]
```diff
@@ -5,6 +5,54 @@
...
+class ComplianceErrorList(IntEnum):
+    C = 0
+    N = 1
+    T = 0
+    V = 0
+    X = 0
+    U = 0
+    D = 0
+
+class ComplianceStatus(StrEnum):
```
[the remark this answers is not in the document]
Comment (lakhotiaharshit) [reply qcodes/instrument_drivers/Keysight/keysightb1500/constants.py]: done

File: qcodes/instrument_drivers/Keysight/keysightb1500/constants.py [hunk 13]
```diff
@@ -5,6 +5,54 @@
...
+    Z = 'XDATA'
+
+class ComplianceErrorList(IntEnum):
+    C = 0
+    N = 1
+    T = 0
+    V = 0
+    X = 0
+    U = 0
+    D = 0
```
[the remark this answers is not in the document]
Comment (lakhotiaharshit) [reply qcodes/instrument_drivers/Keysight/keysightb1500/constants.py]: done

File: qcodes/instrument_drivers/Keysight/keysightb1500/constants.py [hunk 14]
```diff
@@ -5,6 +5,54 @@
...
+    D = 'CH4'
+    E = 'CH5'
+    F = 'CH6'
+    G = 'CH7'
+    H = 'CH8'
+    I = 'CH9'
+    J = 'CH10'
+    Z = 'XDATA'
+
+class ComplianceErrorList(IntEnum):
```
[the remark this answers is not in the document]
Comment (lakhotiaharshit) [reply qcodes/instrument_drivers/Keysight/keysightb1500/constants.py]: Done

File: qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1517A.py [hunk 15]
```diff
@@ -49,6 +79,42 @@ def __init__(self, parent: 'KeysightB1500', name: Optional[str], slot_nr,
...
+    def _get_sampling_number(self) -> int:
+        if self._timing_parameters['number'] is not None:
+            sample_number = self._timing_parameters['number']
+            return sample_number
+        else:
+            raise Exception('set timing parameters first')
+
+    def _get_time_axis(self) -> np.ndarray:
+        sample_rate = self._timing_parameters['interval']
+        total_time = self. _total_measurement_time
```
[the remark this answers is not in the document]
Comment (lakhotiaharshit) [reply qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1517A.py]: Done

File: qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1517A.py [hunk 16]
```diff
@@ -49,6 +79,42 @@ def __init__(self, parent: 'KeysightB1500', name: Optional[str], slot_nr,
...
+        )
+
+        self.add_parameter(
+            name="sampling_measurement",
+            parameter_class=SamplingMeasurement,
+            vals=Arrays(shape=(self._get_sampling_number,)),
+            setpoints=(self.time_axis,)
+        )
+
+    def _get_sampling_number(self) -> int:
```
[the remark this answers is not in the document]
Comment (lakhotiaharshit) [reply qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1517A.py]: done.

File: qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1517A.py (hunk 2 above)
Comment (lakhotiaharshit) [reply qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1517A.py]: Yes, it is too abstract. The name is changed according to your suggestion.

File: qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1517A.py [hunk 17]
```diff
@@ -1,13 +1,20 @@
...
 from .KeysightB1500_module import B1500Module, parse_spot_measurement_response
 from .message_builder import MessageBuilder
 from . import constants
-from .constants import ModuleKind, ChNr, AAD
+from .constants import ModuleKind, ChNr, AAD, MM
 if TYPE_CHECKING:
     from .KeysightB1500 import KeysightB1500
 
 
+# EDIT: Timing parameter: Put validator
```
[the remark this answers is not in the document]
Comment (lakhotiaharshit) [reply qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1517A.py]: interval validator is enough.

File: qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500_module.py [hunk 18]
```diff
@@ -9,6 +10,47 @@
...
+
+    values_separator = ','
+    data_val = []
+    data_status = []
+    data_channel = []
+    data_datatype = []
+
+    for str_value in raw_data_val.split(values_separator):
+        status = str_value[0]
+        channel_name = constants.ChannelName
```
[the remark this answers is not in the document]
Comment (lakhotiaharshit) [reply qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500_module.py]: Only used in the line below. Not needed in general, hence removed.

File: qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500_SamplingMeasurement.py [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+        check for the status other than "N" (normal) and output the
+        number of data values which were not measured under "N" (normal)
+        status.
+
+        For the list of all the status values and their meaning refer to
+        :class:`constants.ComplianceStatus`.
+
+        """
+
+        if isinstance(self.data.status, list):
```
[the remark this answers is not in the document]
Comment (lakhotiaharshit) [reply qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500_SamplingMeasurement.py]: I have updated it to be more explicit

File: qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500_SamplingMeasurement.py [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+            data = self.data
+            total_count = len(data.status)
+            normal_count = data.status.count('N')
```
[the remark this answers is not in the document]
Comment (lakhotiaharshit) [reply qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500_SamplingMeasurement.py]: Updated with enumeration instance for better readability of the code

File: qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500.py [hunk 21]
```diff
@@ -192,5 +192,24 @@ def use_nplc_for_high_resolution_adc(
...
+
+        Args:
+            n: Number of averaging samples, between 1 and 1023. Default
+                setting is 1. (For more info see Table 4-21.)
+        """
+        self._setup_integration_time(
+            adc_type=constants.AIT.Type.HIGH_SPEED,
+            mode=constants.AIT.Mode.MANUAL,
+            coeff=n
+        )
```
Comment (jenshnielsen) [inline qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500.py]: Is there a supported (non private) way to set this back to automatic?

File: qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500_sampling_measurement.py [hunk 22]
```diff
@@ -0,0 +1,98 @@
+import warnings
+
+import numpy
+
+from qcodes import ParameterWithSetpoints
```
Comment (jenshnielsen) [inline qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500_sampling_measurement.py]: Please don't use the global namespace imports within qcodes.

File: qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500_sampling_measurement.py [hunk 23]
```diff
@@ -0,0 +1,98 @@
...
+    """
+
+    _timeout_response_factor = 10.00
+    # This factor is a bit higher than the ratio between
+    # the measured measurement-time and the calculated measurement
+    # (from the user input). Check :get_raw: method to find its usage.
+
+    def __init__(self, name, **kwargs):
+        super().__init__(name, **kwargs)
+        self.data = FMTResponse(None, None, None, None)
```
Comment (jenshnielsen) [inline qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500_sampling_measurement.py]: Should this be private?

File: qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500_sampling_measurement.py [hunk 24]
```diff
@@ -0,0 +1,98 @@
...
+
+    def get_raw(self):
+        """
+        Sets up the measurement by calling :_set_up:
```
Comment (jenshnielsen) [inline qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500_sampling_measurement.py]: I don't think the docs should refer to private methods. Also I don't think `:something:` is valid markup.
I think this should read perform measurement ...

File: qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1517A.py [hunk 25]
```diff
@@ -25,17 +30,40 @@ class B1517A(B1500Module):
...
+                
+                Refer to the documentation of ``MM`` command in the 
+                programming guide for more information.""")
+        )
+        # Instrument is initialized with this setting having value of
+        # `1`, spot measurement mode, hence let's set the parameter to this
+        # value since it is not possible to request this value from the
+        # instrument.
+        self.measurement_mode.raw_value = MM.Mode.SPOT
+        self.measurement_mode._save_val(MM.Mode.SPOT)
```
Comment (jenshnielsen) [inline qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1517A.py]: Why not just pass the initial value to the constructor? It takes initial values for this reason

File: qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1517A.py [hunk 26]
```diff
@@ -145,6 +216,59 @@ def measure_config(self, measure_range: constants.MeasureRange) -> None:
...
+                channels-1)``
+            number: Number of samples. Integer expression. 1 to the
+                following value. Initial value is 1000. For the linear
+                sampling: ``100001 / (number of measurement channels)``.
+                For the log sampling: ``1 + (number of data for 11 decades)``
+            h_base: Hold time of the base value output until the bias value
+                output. Numeric expression. in seconds. 0 (initial setting)
+                to 655.35 s, resolution 0.01 s.
+        """
+        # The diplication of kwargs in the calls below is due to the
```
Comment (jenshnielsen) [inline qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1517A.py]: duplication?

File: qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1517A.py [hunk 27]
```diff
@@ -145,6 +216,59 @@ def measure_config(self, measure_range: constants.MeasureRange) -> None:
...
+        # different in type annotations between ``MessageBuilder.mt()``
+        # method and ``_timing_parameters`` attribute.
+
+        IntervalValidator = vals.Numbers(0.0001, 65.535)
+        IntervalValidator.validate(interval)
```
Comment (jenshnielsen) [inline qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1517A.py]: Should this validator be a private attribute on the class so that it is only created once?

File: qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500_sampling_measurement.py (hunk 22 above)
Comment (lakhotiaharshit) [reply qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500_sampling_measurement.py]: Lazy mistake, taken care of

File: qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500.py (hunk 21 above)
Comment (lakhotiaharshit) [reply qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500.py]: At the moment no. However, if this issue crops up (some user requests for it) we can make _setup_integration_time public.

File: qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500_sampling_measurement.py (hunk 23 above)
Comment (lakhotiaharshit) [reply qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500_sampling_measurement.py]: Yes, Thanks !

File: qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500_sampling_measurement.py (hunk 24 above)
Comment (lakhotiaharshit) [reply qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500_sampling_measurement.py]: Right. This docstring has been reworked.

File: qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1517A.py (hunk 25 above)
Comment (astafan8) [reply qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1517A.py]: passing the `initial_value` will call `set` with it on the parameter, which i decided to be undesirable some weeks ago, but now i don't remember why i decided this :( perhaps i wanted to avoid 'setting' things at "init"...

what would be your recommendation here?

File: qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1517A.py (hunk 26 above)
Comment (lakhotiaharshit) [reply qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1517A.py]: Corrected

File: qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1517A.py (hunk 27 above)
Comment (lakhotiaharshit) [reply qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1517A.py]: Yes. Thanks for pointing it out. It is implemented.

File: qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1517A.py (hunk 25 above)
Comment (jenshnielsen) [reply qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1517A.py]: I don't like calling private methods but perhaps we just need to extend inital_value to also handle this corner case. No strong objection against leave this as is

File: qcodes/tests/drivers/keysight_b1500/b1500_driver_tests/test_sampling_measurement.py [hunk 28]
```diff
@@ -0,0 +1,116 @@
...
+    smu_sampling_measurement.timing_parameters(h_bias=0,
+                                               interval=0.1,
+                                               number=n_samples)
+    smu_sampling_measurement.sampling_measurement.get()
+    compliance_list_string = [status]*n_samples
+    compliance_list = [constants.ComplianceErrorList[i[0]].value
+                       for i in compliance_list_string]
+
+    assert isinstance(
+        smu_sampling_measurement.sampling_measurement.compliance(), list)
```
[the remark this answers is not in the document]
Comment (lakhotiaharshit) [reply qcodes/tests/drivers/keysight_b1500/b1500_driver_tests/test_sampling_measurement.py]: Thanks! resolved

File: qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500_SamplingMeasurement.py [hunk 29]
```diff
@@ -0,0 +1,102 @@
...
+        time_out = measurement_time * self._timeout_response_factor
+        default_timeout = self.root_instrument.timeout()
+
+        # if time out to be set is lower than the default value
+        # then keep default
+        if time_out < default_timeout:
+            time_out = default_timeout
+
+        with self.root_instrument.timeout.set_to(time_out):
+            self._set_up()
```
[the remark this answers is not in the document]
Comment (lakhotiaharshit) [reply qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500_SamplingMeasurement.py]: Thanks! done.

File: qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500_sampling_measurement.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (lakhotiaharshit) [reply qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500_sampling_measurement.py]: done

File: qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500_module.py [hunk 30]
```diff
@@ -9,6 +10,42 @@
     from .KeysightB1500 import KeysightB1500
 
 
+FMTResponse = namedtuple('FMTResponse', 'value status channel type')
+
+
+def parse_fmt_1_0_response(raw_data_val):
```
[the remark this answers is not in the document]
Comment (lakhotiaharshit) [reply qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500_module.py]: Done

File: qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1517A.py [hunk 31]
```diff
@@ -49,6 +77,43 @@ def __init__(self, parent: 'KeysightB1500', name: Optional[str], slot_nr,
...
 
+        self.add_parameter(
+            name="time_axis",
+            get_cmd=self._get_time_axis,
+            vals=Arrays(shape=(self._get_number_of_samples,)),
+            snapshot_value=False
+        )
+
+        self.add_parameter(
+            name="sampling_measurement_trace",
```
[the remark this answers is not in the document]
Comment (lakhotiaharshit) [reply qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1517A.py]: We resolved this together by setting the labels and unit explicitly.

File: qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1517A.py [hunk 32]
```diff
@@ -49,6 +77,43 @@ def __init__(self, parent: 'KeysightB1500', name: Optional[str], slot_nr,
             get_cmd=self._get_current
         )
 
+        self.add_parameter(
+            name="time_axis",
```
[the remark this answers is not in the document]
Comment (lakhotiaharshit) [reply qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1517A.py]: added
